### PR TITLE
Add CStringArray::join() method

### PR DESCRIPTION
**Pull request by mikee47**
Available in other languages (python, javascript) this provides an efficient way to concatenate array members.

### CODE REVIEW

File: Sming/Core/Data/CStringArray.h [hunk 1]
```diff
@@ -246,6 +246,15 @@ class CStringArray : private String
...
 	unsigned count() const;
 
+	/**
+	 * @brief Get contents of array as delimited string
+	 * @param separator What to join elements with
+	 * @retval String
+	 *
+	 * e.g. CStringArray(F("a\0b\0c")).join() returns "a,b,c"
+	 */
+	String join(const String& separator = ",") const;
```
Comment (slaff) [inline Sming/Core/Data/CStringArray.h:256]: @mikee47 Is there a reason for the separator parameter to be optional? In JavaScript, PHP and Python as far as I can say this parameter is mandatory.

Comment (mikee47) [reply Sming/Core/Data/CStringArray.h:256]: I hadn't considered PHP but default is same as javascript https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join.

Python would be `",".join(cs)` which is an interesting construction - clearly there cannot be any 'default' in this case. However, to actually implement this efficiently for Sming would require that the array itself does the join. We could add this as a template method to String easily enough:

```
class String
{
  ...

  template <class Array> String join(const Array& array)
  {
    return array.join(*this);
  }

  ...
};
```

Comment (mikee47) [reply Sming/Core/Data/CStringArray.h:256]: I don't know of any C++ STL equivalent for this, though I see Boost has https://www.geeksforgeeks.org/boostalgorithmjoin-in-cpp-library/ https://www.boost.org/doc/libs/1_83_0/doc/html/boost/algorithm/join.html

Comment (mikee47) [reply Sming/Core/Data/CStringArray.h:256]: > to actually implement this efficiently for Sming would require that the array itself does the join

The efficiency here is avoiding reallocations for the returned String as in the general case:

```
template <class Array> String join(const Array& array, const String& separator)
{
  String s;
  for(auto& e: array) {
    if(s) {
      s += separator;
    }
    s += e;
  }
  return s;
}
```

This should work for any `Vector` or other iterable objects with an implicit String() operator. Something that works with HashMaps would also be useful (returning delimited 'name=value' pairs). However the implementation as it stands does not perform any string escaping, for example:

```
CStringArray cs;
cs.add("first");
cs.add("second,plus,stuff");
```
Then `cs.join()` returns "first,second,plus,stuff" as requested. If "first,\"second,plus,stuff\"" is required that'll have to be done manually. This is same behaviour as python.

**Pull request merged**